### PR TITLE
Specs under registries

### PR DESCRIPTION
**Pull request by djwhatle**
- Move specs to reside under registries in the config file.
 - Add "REGISTRY" header to `sbcli list`
 - `sbcli list` will automatically retrieve new specs for any empty registries found in the config file
 - Add some printing functionality

### CODE REVIEW

**Comment by djwhatle** [comment]
Did a bit more cleanup on printing as part of this. Looks like this now.

```
[dwhatley@precision-t sbcli]$ ./sbcli registry list
Found registries already in config:
 NAME          TYPE           ORG                        URL        
 --------  -+- ---------  -+- ---------------------  -+- ---------  
 examples   |  dockerhub   |  ansibleplaybookbundle   |  docker.io  
 aws        |  dockerhub   |  awsservicebroker        |  docker.io  
```

```
[dwhatley@precision-t sbcli]$ ./sbcli bundle list
Found specs already in registry: [examples]
Found specs already in registry: [aws]
 BUNDLE                     IMAGE                                                                    REGISTRY  
 ---------------------  -+- -------------------------------------------------------------------  -+- --------  
 hastebin-apb            |  docker.io/ansibleplaybookbundle/hastebin-apb:latest                   |  examples  
 hello-world-apb         |  docker.io/ansibleplaybookbundle/hello-world-apb:latest                |  examples  
 demo-api-apb            |  docker.io/ansibleplaybookbundle/photo-album-demo-api-apb:latest       |  examples  
 manageiq-apb            |  docker.io/ansibleplaybookbundle/manageiq-apb:latest                   |  examples  
 import-vm-apb           |  docker.io/ansibleplaybookbundle/import-vm-apb:latest                  |  examples  
 prometheus-apb          |  docker.io/ansibleplaybookbundle/prometheus-apb:latest                 |  examples  
 v2v                     |  docker.io/ansibleplaybookbundle/v2v-apb:latest                        |  examples  
 import-vm-disk-apb      |  docker.io/ansibleplaybookbundle/import-vm-disk-apb:latest             |  examples  
 pyzip-demo-apb          |  docker.io/ansibleplaybookbundle/pyzip-demo-apb:latest                 |  examples  
 rekognition             |  docker.io/awsservicebroker/rekognition-apb:latest                     |  aws       
 route53                 |  docker.io/awsservicebroker/route53-apb:latest                         |  aws       
 translate               |  docker.io/awsservicebroker/translate-apb:latest                       |  aws

File: cmd/bundle.go [hunk 1]
```diff
@@ -84,80 +84,78 @@ func init() {
...
+// Get images from a single registry
+func getImages(registryMetadata Registry) ([]*bundle.Spec, error) {
 	var specList []*bundle.Spec
-	err := viper.UnmarshalKey("Registries", &regConfigList)
+
+	authNamespace := ""
+	registry, err := registries.NewRegistry(registryMetadata.Config, authNamespace)
 	if err != nil {
-		log.Error("Error unmarshalling config: ", err)
+		log.Error("Error from creating a NewRegistry")
```
Comment (dymurray) [inline cmd/bundle.go]: Make this `log.Errorf("Error from creating a NewRegistry - %v", err)`

Comment (djwhatle) [reply cmd/bundle.go]: I must have messed this up in one of my rebases, don't know why I would have removed an error print.

File: cmd/bundle.go [hunk 2]
```diff
@@ -84,80 +84,78 @@ func init() {
...
+	for _, regConfig := range regConfigs {
+		if len(regConfig.Specs) > 0 && Refresh == false {
+			fmt.Printf("Found specs already in registry: [%s]\n", regConfig.Config.Name)
+			newRegConfigs = append(newRegConfigs, regConfig)
+			continue
+		}
+		fmt.Printf("Getting specs for registry: [%s]\n", regConfig.Config.Name)
+		specs, err := getImages(regConfig)
+		if err != nil {
+			log.Error("Error getting images")
```
[the remark this answers is not in the document]
Comment (djwhatle) [reply cmd/bundle.go]: Same as below, must have lost this in a rebase. Good catch!

**Comment by djwhatle** [comment]
@dymurray added the changes you suggested.

**Comment by jmrodri** [comment]
First time I ran it I got the following:
```
[jesusr@speed3 sbcli{djwhatle-specs-under-registries}]$ make
go build -i -ldflags="-s -w"
[jesusr@speed3 sbcli{djwhatle-specs-under-registries}]$ ./sbcli registry list
Found registries already in config:
 NAME      TYPE      ORG      URL  
   -+-   -+-   -+-   
    |     |     |    
```
Then I ran `bundle list`
```
[jesusr@speed3 sbcli{djwhatle-specs-under-registries}]$ ./sbcli bundle list
Getting specs for registry: []
ERRO Error from creating a NewRegistry            
ERRO unable to validate registry name             
ERRO Error getting images - unable to validate registry name 
 BUNDLE      IMAGE      REGISTRY  
   -+-   -+-   
```

Another `registry list`
```
[jesusr@speed3 sbcli{djwhatle-specs-under-registries}]$ ./sbcli registry list
Found no registries in configuration. Try `sbcli registry add`.
```

**Comment by jmrodri** [comment]
The problem is using an older `.sbcli.json` file causes it not to parse the registries correctly reporting empty data. 

### OLDD format
```json
  "registries": [
    {
      "AuthName": "",
      "AuthType": "",
      "BlackList": null,
      "Fail": false,
      "Images": null,
      "Name": "zeus",
      "Namespaces": null,
      "Org": "jmrodri",
      "Pass": "",
      "Runner": "",
      "Tag": "",
      "Type": "dockerhub",
      "URL": "docker.io",
      "User": "",
      "WhiteList": [
        ".*-apb$"
      ]
    }
  ],
```

### NEW format
```json
  "registries": [
    {
      "Config": {
        "URL": "docker.io",
        "AuthType": "",
        "AuthName": "",
        "User": "",
        "Pass": "",
        "Org": "ansibleplaybookbundle",
        "Tag": "",
        "Type": "dockerhub",
        "Name": "docker",
        "Runner": "",
        "Images": null,
        "Namespaces": null,
        "Fail": false,
        "WhiteList": [
          ".*-apb$"
        ],
        "BlackList": null
      },
      "Specs": null
    }
  ],
```

If you see output like the one below, it is very likely you have an old config, delete it and recreate it.

```
$ ./sbcli registry list
Found registries already in config:
 NAME      TYPE      ORG      URL  
   -+-   -+-   -+-   
    |     |     |    
```

**Comment by djwhatle** [comment]
@jmrodri good call on documented the sbcli.json format change here. I should have included that from the beginning.

That output on sbcli registry list doesn't look like what I would expect for an empty list. I think I lost a small change I made to printing somewhere along the way. Will make a new (very small) PR with that fix.

**Pull request merged**